apps/emqx_retainer/src/emqx_retainer_index.erl

```diff
@@ -0,0 +1,144 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2020-2022 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+-module(emqx_retainer_index).
+
+-export([foreach_index_key/3,
+         convert_tokens_for_index/2,
+         index_score/2,
+         select_index/2,
+         condition/1,
+         condition/2,
+         restore_topic/3
+        ]).
+
+-export_type([index/0]).
+
+-type index() :: list(pos_integer()).
+
+-type index_key() :: {index(), {emqx_topic:words(), emqx_topic:words()}}.
+
+%%--------------------------------------------------------------------
+%% API
+%%--------------------------------------------------------------------
+
+-spec foreach_index_key(fun((index_key()) -> any()), list(index()), emqx_topic:words()) -> ok.
+foreach_index_key(_Fun, [], _Tokens) -> ok;
+
+foreach_index_key(Fun, [Index | Indexes], Tokens) ->
```

Comment:
Nit, since we've decided to lean towards British English (e.g. `behaviour` vs. `behavior`), it's more consistent to use `Indices` as plural of `Index`.